### PR TITLE
Switch to Debian12

### PR DESCRIPTION
**Pull request by gilesknap**
This is marked as draft because it is alternative to and replaces the other two PRs I made.

This adds a vagrant build based on official Debian Bookworm and switches ansible to use DEBs instead of RPMs.

Just now working on the DEB/RPM agnostic version ...

I have squashed the commits down to two:
- Adding Vagrant with Fedora and minimal changes to ansible
- Switching to Debian

This means we have clear starting points to try any DEB or RPM distro.

To test this PR.
- clone this repo
- follow https://github.com/gilesknap/training-vm/blob/vagrant-debian/doc/vagrant.md
- Once the VM is up:-
```bash
VM_REPO="https://github.com/gilesknap/training-vm.git -b debian12" eval "$(curl -L https://raw.githubusercontent.com/gilesknap/training-vm/debian12/bootstrap.sh)"

vim training/local.yml # uncomment choice of roles and modules
./bootstrap/update.sh training
```

I have all the roles and modules building except for pvapy which has some challenges with boost. I'm sure that can be fixed easily enough.

Note that this uses docker rootful instead of podman rootless. Because: https://github.com/epics-training/training-vm/pull/22#issuecomment-2370878140

### CODE REVIEW

**Comment by gilesknap** [comment]
tagging those that might like to review.
@ralphlange
@mrakitin
@shroffk
@MJGaughran
@marcelldls
@coretl

File: ansible/roles/epics-modules/tasks/build_module.yml [hunk 1]
```diff
@@ -54,26 +53,29 @@
...
+    regexp: "^SUPPORT="
   become: true
 
 - name: "{{ epics_modules[item].name }} | Run pre-build hook"
   include_tasks: "{{ epics_modules[item].pre_hook }}"
   when: epics_modules[item].pre_hook is defined
 
 - name: "{{ epics_modules[item].name }} | Build from sources (this may take a while)"
-  command: make -C "{{ epics_install_path }}/{{ item }}-{{ epics_modules[item].version }}" all clean
+  command: make {{ epics_modules[item]. args | default('') }} -C "{{ epics_install_path }}/{{ item }}-{{ epics_modules[item].version }}" all clean
```
Comment (ralphlange) [inline ansible/roles/epics-modules/tasks/build_module.yml:72]: `epics_modules[item]. args` That space looks suspicious.

Comment (gilesknap) [reply ansible/roles/epics-modules/tasks/build_module.yml:72]: for sure. I wonder why it works?
image

Comment (ralphlange) [reply ansible/roles/epics-modules/tasks/build_module.yml:72]: Probably the parser is very forgiving.

**Comment by ralphlange** [comment]
At least for the EPICS modules, one of the training-vm's really useful take-home goodies is the list of build dependencies (package names) for each module.

While I completely understand your suggestion to restrict the target systems to save effort, that information would be lost for the RedHat world - currently the major part of our users.

**Comment by gilesknap** [comment]
I see. In that case we need to merge the two. Each time we update we need to check against (at least) two distros.

But with a 12 minute build that is not too hard.

This inspires me to run the user mode ansible automatically from vagrant after the with all of the roles and modules enabled. Then we can put that in CI.

**Comment by gilesknap** [comment]
> I see. In that case we need to merge the two. Each time we update we need to check against (at least) two distros.
> 
> But with a 12 minute build that is not too hard.
> 
> This inspires me to run the user mode ansible automatically from vagrant after the initial build with all of the roles and modules enabled. Then we can put that in CI. And we can snapshot the initial build for delivery as an OVA.

:-) https://github.com/marketplace/actions/vagrant-virtualbox-with-repos-vagrantfile

**Comment by ralphlange** [comment]
Yesss.
Adding CI builds was what I had in mind.

**Comment by gilesknap** [comment]
I'm happy to have a go at doing the merge and CI. Maybe in a couple of weeks.
It's not going to be quite so quick on a github runner :-)

**Comment by gilesknap** [comment]
@ralphlange - now I'm thinking we should always do all the ansible in the guest.
- reduces the installs on the host
- will work for windows and linux

We can just provision ansible with a shell provisioner and then do the rest with ansible.

**Comment by ralphlange** [comment]
That's what I did (still on the Fedora level of your PR) and it works fine.

```
...
    local.vm.synced_folder "../ansible", "/ansible"
...
    # Install and run guest-side Ansible
    local.vm.provision "shell", inline: <<-SHELL
      sudo dnf -y install ansible
      ansible-galaxy install -r /ansible/requirements.yml || true
      ansible-playbook -i /ansible/hosts -t initial_setup /ansible/playbook.yml
    SHELL
...
```

**Comment by gilesknap** [comment]
I may have overestimated how hard the merge was. Ansible is great at this. Will make another PR today.

I'm also beginning to think that epics containers should share this file https://github.com/epics-training/training-vm/blob/main/ansible/roles/epics-modules/vars/main.yml and adopt anisble instead of `ibek` for the build stage.
@coretl - do you have appetite for one more major breaking change? It would be a chance to throw away more bespoke code (and ibek would come full circle to your original concept).

**Comment by ralphlange** [comment]
I'm getting scared... 🤨
But, yes: sounds like a choice worth exploring.

**Comment by gilesknap** [comment]
> I'm getting scared... 🤨 But, yes: sounds like a choice worth exploring.

You are solving many of the same issues I have had to solve in building IOC containers. But you used a standard, very popular tool to do so. I have broken my 'Principal Of Minimum Bespoke' by writing my own python to solve problems that Ansible already did.

**Comment by ralphlange** [comment]
On Fedora 40 guest (but I had the same behavior with - manually created - Rocky guests):
Shared Clipboard and Drag-and-Drop between host and guest only work when the guest is running a "GNOME on Xorg" session, i.e., using X instead of Wayland.

Unless you or @shroffk see graphics performance issues, I would suggest making the X session the default. (If that is possible.)

**Comment by gilesknap** [comment]
> Unless you or @shroffk see graphics performance issues, I would suggest making the X session the default. (If that is possible.)

That seems sensible to me. I had not noticed that it was using wayland - but I had loads of issues with wayland and ubuntu guest on windows host in VirtualBox

**Comment by gilesknap** [comment]
> ansible-galaxy install -r /ansible/requirements.yml || true

@ralphlange please can you share your requirements.yml.

**Comment by gilesknap** [comment]
We now have DEB/RPM agnostic ansible and vagrant supporting Debian or Fedora https://github.com/epics-training/training-vm/pull/27.

Also:
- all ansible runs in the guest
- attempted to enable Xorg only but its not working on Debian.

**Comment by gilesknap** [comment]
Closing this and moving over to https://github.com/epics-training/training-vm/pull/27

**Pull request closed without merging**